source/base/hdf5.cc

```diff
@@ -0,0 +1,1752 @@
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &*data.begin();
+    }
+
+
+
+    template <typename number>
+    const void *
+    get_container_const_pointer(const FullMatrix<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0][0];
+    }
+
+
+
+    // This function initializes and returns a container of type std::vector,
+    // Vector or FullMatrix. The function does not set the values of the
+    // elements of the container. The container can store data of a HDF5 dataset
+    // or a HDF5 selection. The dimensions parameter holds the dimensions of the
+    // HDF5 dataset or selection.
+    //
+    // In the case of a std::vector, the size of the vector will be the total
+    // size given by dimensions. For example in the case of a dataset of rank 3,
+    // the dimensions are std::vector<hsize_t>{dim_0,dim_1,dim_2}. The size of
+    // the returned std::vector will be dim_0*dim_1*dim_2
+    //
+    // In the case of a dealii::Vector, the size of the returned dealii::Vector
+    // will be as well dim_0*dim_1*dim_2
+    //
+    // A FullMatrix can store only data of HDF5 datasets with rank 2. The size
+    // of the FullMatrix will be FullMatrix(dim_0,dim_2)
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container,
+                   std::vector<typename Container::value_type>>::value,
+      Container>::type
+    initialize_container(std::vector<hsize_t> dimensions)
+    {
+      return Container(std::accumulate(
+        dimensions.begin(), dimensions.end(), 1, std::multiplies<int>()));
+    }
+
+
+
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container, Vector<typename Container::value_type>>::value,
+      Container>::type
+    initialize_container(std::vector<hsize_t> dimensions)
+    {
+      return Container(std::accumulate(
+        dimensions.begin(), dimensions.end(), 1, std::multiplies<int>()));
+    }
+
+
+
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container,
+                   FullMatrix<typename Container::value_type>>::value,
+      Container>::type
+    initialize_container(std::vector<hsize_t> dimensions)
+    {
+      // If the rank is higher than 2, then remove single-dimensional entries
+      // from the shape defined by dimensions. This is equivalent to the squeeze
+      // function of python/numpy. For example the following code would convert
+      // the vector {1,3,1,2} to {3,2}
+      std::vector<hsize_t> squeezed_dimensions;
+
+      if (dimensions.size() > 2)
+        {
+          for (auto &&dimension : dimensions)
```

Comment:
I think that if you make `dimensions` a const reference, you might have to make the type of `dimension` a plain reference, not an rvalue reference.